:orphan:

:py:mod:`trieste.utils.misc`
============================

.. py:module:: trieste.utils.misc


Module Contents
---------------

.. py:data:: C
   

   A type variable bound to `typing.Callable`. 


.. py:function:: jit(apply: bool = True, **optimize_kwargs: Any) -> Callable[[C], C]

   A decorator that conditionally wraps a function with `tf.function`.

   :param apply: If `True`, the decorator is equivalent to `tf.function`. If `False`, the decorator
       does nothing.
   :param optimize_kwargs: Additional arguments to `tf.function`.
   :return: The decorator.


.. py:function:: shapes_equal(this: trieste.types.TensorType, that: trieste.types.TensorType) -> trieste.types.TensorType

   Return a scalar tensor containing: `True` if ``this`` and ``that`` have equal runtime shapes,
   else `False`.


.. py:function:: to_numpy(t: trieste.types.TensorType) -> numpy.ndarray

   :param t: An array-like object.
   :return: ``t`` as a NumPy array.


.. py:data:: T_co
   

   An unbounded covariant type variable. 


.. py:class:: Result

   Bases: :py:obj:`Generic`\ [\ :py:obj:`T_co`\ ], :py:obj:`abc.ABC`

   Represents the result of an operation that can fail with an exception. It contains either the
   operation return value (in an :class:`Ok`), or the exception raised (in an :class:`Err`).

   To check whether instances such as

       >>> res = Ok(1)
       >>> other_res = Err(ValueError("whoops"))

   contain a value, use :attr:`is_ok` (or :attr:`is_err`)

       >>> res.is_ok
       True
       >>> other_res.is_ok
       False

   We can access the value if it :attr:`is_ok` using :meth:`unwrap`.

       >>> res.unwrap()
       1

   Trying to access the value of a failed :class:`Result`, or :class:`Err`, will raise the wrapped
   exception

       >>> other_res.unwrap()
       Traceback (most recent call last):
           ...
       ValueError: whoops

   **Note:** This class is not intended to be subclassed other than by :class:`Ok` and
   :class:`Err`.

   .. py:method:: is_ok(self) -> bool
      :property:

      `True` if this :class:`Result` contains a value, else `False`.


   .. py:method:: is_err(self) -> bool
      :property:

      `True` if this :class:`Result` contains an error, else `False`. The opposite of
      :attr:`is_ok`.


   .. py:method:: unwrap(self) -> T_co
      :abstractmethod:

      :return: The contained value, if it exists.
      :raise Exception: If there is no contained value.



.. py:class:: Ok(value: T_co)

   Bases: :py:obj:`Result`\ [\ :py:obj:`T_co`\ ]

   Wraps the result of a successful evaluation.

   :param value: The result of a successful evaluation.

   .. py:method:: is_ok(self) -> bool
      :property:

      `True` always.


   .. py:method:: unwrap(self) -> T_co

      :return: The wrapped value.



.. py:class:: Err(exc: Exception)

   Bases: :py:obj:`Result`\ [\ :py:obj:`NoReturn`\ ]

   Wraps the exception that occurred during a failed evaluation.

   :param exc: The exception that occurred.

   .. py:method:: is_ok(self) -> bool
      :property:

      `False` always.


   .. py:method:: unwrap(self) -> NoReturn

      :raise Exception: Always. Raises the wrapped exception.



.. py:class:: DEFAULTS

   Default constants used in Trieste.

   .. py:attribute:: JITTER
      :annotation: :typing_extensions.Final[float] = 1e-06

      The default jitter, typically used to stabilise computations near singular points, such as in
      Cholesky decomposition.



